src/lib.rs

```diff
@@ -159,33 +159,34 @@ pub type LevelRun = Range<usize>;
 /// `line` is a range of bytes indices within `paragraph`.
 ///
 /// http://www.unicode.org/reports/tr9/#Reordering_Resolved_Levels
-pub fn visual_runs(line: Range<usize>,
-                   para_level: u8,
-                   max_level: u8,
-                   levels: &[u8]) -> Vec<LevelRun> {
-    assert!(line.start <= levels.len());
-    assert!(line.end <= levels.len());
+pub fn visual_runs(line: Range<usize>, info: &ParagraphInfo) -> Vec<LevelRun> {
+    assert!(line.start <= line.end);
+    assert!(line.end <= info.levels.len());
 
     // TODO: Whitespace handling.
     // http://www.unicode.org/reports/tr9/#L1
 
-    assert!(max_level >= para_level);
-    let mut runs = Vec::with_capacity((max_level - para_level) as usize + 1);
+    assert!(info.max_level >= info.para_level);
+    let mut runs = Vec::with_capacity((info.max_level - info.para_level) as usize + 1);
 
     // Optimization: If there's only one level, just return a single run for the whole line.
-    if max_level == para_level || line.len() == 0 {
+    if info.max_level == info.para_level || line.len() == 0 {
         runs.push(line.clone());
         return runs
     }
 
     // Find consecutive level runs.
     let mut start = line.start;
-    let mut level = levels[start];
-    let mut min_level = level;
-    let mut max_level = level;
+    let mut level = info.levels[start];
+    let mut min_level = level;  // FIXME: Is this the same as info.max_level?
+    let mut max_level = level;  // FIXME: Is this the same as info.para_level?
 
+    // FIXME: Could this loop be replaced with a call to level_runs()?
```

Comment:
Using `info.{para_level, max_level}` would be correct but potentially sub-optimal, since that range might not occur in the given line.  We could make `level_runs` return the min and max levels, though this would be wasted work when it's called from `isolating_run_sequences`...